File: quiz-engine-golang/internal/engine/quiz.go
package engine

import (
	"database/sql"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/pbaletkeman/quiz-engine-golang/internal/database"
	"github.com/pbaletkeman/quiz-engine-golang/internal/models"
)

// QuizConfig holds configuration for a quiz session.
type QuizConfig struct {
	NumQuestions int
	Difficulty   string
	Section      string
	DBPath       string
}

// QuizEngine manages a single quiz session.
type QuizEngine struct {
	SessionID   string
	Questions   []models.Question
	Responses   []models.QuizResponse
	NumCorrect  int
	DB          *sql.DB
	Config      QuizConfig
	StartedAt   time.Time
	ShuffleData []ShuffleResult
}

// NewQuizEngine creates a new QuizEngine with a fresh session ID.
func NewQuizEngine(db *sql.DB, config QuizConfig) *QuizEngine {
	return &QuizEngine{
		SessionID: uuid.New().String(),
		DB:        db,
		Config:    config,
		StartedAt: time.Now(),
	}
}

// LoadQuestions fetches random questions and creates the session record.
func (qe *QuizEngine) LoadQuestions() error {
	questions, err := database.GetRandomQuestions(qe.DB, qe.Config.NumQuestions, qe.Config.Difficulty, qe.Config.Section)
	if err != nil {
		return err
	}
	qe.Questions = questions

	qe.ShuffleData = make([]ShuffleResult, len(questions))
	for i, q := range questions {
		fullQ, err := database.GetQuestionWithAnswer(qe.DB, q.ID)
		if err == nil && fullQ != nil {
			qe.Questions[i].CorrectAnswer = fullQ.CorrectAnswer
		}
		options := []string{qe.Questions[i].OptionA, qe.Questions[i].OptionB, qe.Questions[i].OptionC, qe.Questions[i].OptionD}
		if qe.Questions[i].OptionE != "" {
			options = append(options, qe.Questions[i].OptionE)
		}
		qe.ShuffleData[i] = ShuffleAnswers(options, qe.Questions[i].CorrectAnswer)
	}

	session := models.QuizSession{
		SessionID:    qe.SessionID,
		StartedAt:    qe.StartedAt,
		NumQuestions: len(questions),
	}
	return database.CreateSession(qe.DB, session)
}

// SubmitAnswer records the user's answer for question at qIdx.
func (qe *QuizEngine) SubmitAnswer(qIdx int, userAnswer string, timeTaken int) error {
	if qIdx >= len(qe.Questions) {
		return fmt.Errorf("question index out of range")
	}
	q := qe.Questions[qIdx]
	shuffle := qe.ShuffleData[qIdx]

	userAnswerIdx := map[string]int{"A": 0, "B": 1, "C": 2, "D": 3, "E": 4}
	originalLetter := ""
	if idx, ok := userAnswerIdx[userAnswer]; ok && idx < len(shuffle.ShuffledOptions) {
		originalLetter = shuffle.PositionMap[idx]
	}

	isCorrect := originalLetter == q.CorrectAnswer
	if isCorrect {
		qe.NumCorrect++
	}

	resp := models.QuizResponse{
		SessionID:        qe.SessionID,
		QuestionID:       q.ID,
		UserAnswer:       userAnswer,
		IsCorrect:        isCorrect,
		TimeTakenSeconds: timeTaken,
	}
	qe.Responses = append(qe.Responses, resp)
	return database.SaveResponse(qe.DB, resp)
}

// FinalizeQuiz closes the session and advances the question cycle if needed.
func (qe *QuizEngine) FinalizeQuiz() (models.QuizSession, error) {
	now := time.Now()
	timeTaken := int(now.Sub(qe.StartedAt).Seconds())
	pct := 0.0
	if len(qe.Questions) > 0 {
		pct = float64(qe.NumCorrect) / float64(len(qe.Questions)) * 100
	}

	session := models.QuizSession{
		SessionID:         qe.SessionID,
		StartedAt:         qe.StartedAt,
		EndedAt:           &now,
		NumQuestions:      len(qe.Questions),
		NumCorrect:        qe.NumCorrect,
		PercentageCorrect: pct,
		TimeTakenSeconds:  timeTaken,
	}

	err := database.UpdateSession(qe.DB, session)
	if err != nil {
		return session, err
	}

	for _, q := range qe.Questions {
		database.MarkQuestionUsed(qe.DB, q.ID)
	}
	database.AdvanceCycleIfExhausted(qe.DB)

	return session, nil
}

// GetReviewData returns full questions (with answers) and the responses from this session.
func (qe *QuizEngine) GetReviewData() ([]models.Question, []models.QuizResponse, error) {
	fullQuestions := make([]models.Question, len(qe.Questions))
	for i, q := range qe.Questions {
		fq, err := database.GetQuestionWithAnswer(qe.DB, q.ID)
		if err != nil || fq == nil {
			fullQuestions[i] = q
		} else {
			fullQuestions[i] = *fq
		}
	}
	return fullQuestions, qe.Responses, nil
}

File: quiz-python-engine/quiz_engine/quiz.py
from typing import List, Optional
import uuid
from datetime import datetime
from .database import DatabaseManager
from .models import Question, QuizSession, QuizResponse
from .utils import calculate_score


class QuizEngine:
    def __init__(self, db: DatabaseManager, num_questions: int = 100,
                 difficulty: str = None, section: str = None):
        self.db = db
        self.num_questions = num_questions
        self.difficulty = difficulty
        self.section = section
        self.session_id = str(uuid.uuid4())
        self.questions: List[Question] = []
        self._correct_answers: dict = {}  # question_id -> correct_answer (private)
        self.responses: List[QuizResponse] = []
        self.num_correct = 0
        self.started_at = datetime.utcnow()

    def load_questions(self):
        # Load display questions (no correct_answer)
        self.questions = self.db.get_random_questions(
            self.num_questions, self.difficulty, self.section
        )
        # Separately fetch correct answers for internal checking
        question_ids = [q.id for q in self.questions if q.id is not None]
        if question_ids:
            questions_with_answers = self.db.get_questions_with_answers(question_ids)
            self._correct_answers = {q.id: q.correct_answer for q in questions_with_answers}

    def submit_answer(self, question_idx: int, user_answer: str, time_taken: int = 0):
        question = self.questions[question_idx]
        correct = self._correct_answers.get(question.id)
        is_correct = (user_answer.upper() == correct) if correct else False

        if is_correct:
            self.num_correct += 1

        response = QuizResponse(
            session_id=self.session_id,
            question_id=question.id,
            user_answer=user_answer.upper(),
            is_correct=is_correct,
            time_taken_seconds=time_taken,
        )
        self.responses.append(response)
        self.db.insert_response(response)
        return is_correct

    def finalize(self) -> QuizSession:
        ended_at = datetime.utcnow()
        time_taken = int((ended_at - self.started_at).total_seconds())
        percentage = calculate_score(self.num_correct, len(self.questions))

        session = QuizSession(
            session_id=self.session_id,
            started_at=self.started_at,
            ended_at=ended_at,
            num_questions=len(self.questions),
            num_correct=self.num_correct,
            percentage_correct=percentage,
            time_taken_seconds=time_taken,
        )
        self.db.insert_session(session)

        # Mark questions as used
        for q in self.questions:
            if q.id is not None:
                self.db.mark_question_used(q.id)

        # Auto-advance cycle if exhausted
        self.db.advance_cycle_if_exhausted()

        return session

    def get_results(self) -> QuizSession:
        session = self.db.get_session(self.session_id)
        return session

    def get_session_review(self) -> dict:
        question_ids = [q.id for q in self.questions if q.id is not None]
        questions_with_answers = self.db.get_questions_with_answers(question_ids)
        q_map = {q.id: q for q in questions_with_answers}

        incorrect = []
        correct = []

        for response in self.responses:
            q = q_map.get(response.question_id)
            if q:
                item = {
                    'question': q,
                    'response': response,
                }
                if response.is_correct:
                    correct.append(item)
                else:
                    incorrect.append(item)

        return {'incorrect': incorrect, 'correct': correct}
